torch/testing/_internal/distributed/rpc/dist_autograd_test.py

```diff
@@ -2065,6 +2065,20 @@ def backward(ctx, grad):
             # refcount.
             self.assertTrue(p_g == p_a)
 
+    @dist_init
+    def test_post_hooks(self):
+        linear = torch.nn.Linear(1, 1, bias=False)
+        self.hook_called_times = 0
+        def post_hook(m, grad_input, grad_output):
+            self.hook_called_times += 1
+        # Note: this actually doesn't install a post hook in AccumulateGrad.
```

Comment:
This doesn't really test the codepaths we have modified in this PR right? We should test post hooks on AccumulateGrad itself. An example of how to add post hooks to AccumulateGrad:

```
import torch

count = 0
def post_hook(grad_input, grad_output):
    global count
    count += 1
    return grad_input

t = torch.rand(10, 10, requires_grad=True)
a = t + t
accumulate_grad = a.grad_fn.next_functions[0][0]
accumulate_grad.register_hook(post_hook)
a.sum().backward()
print(count)
```

Lets test both cases: 1) post hooks on non-leaf grad functions 2) post hooks on accumulate grad

Reply:
> 1) post hooks on non-leaf grad functions

We can test that, but want to make sure we are on the same page. Post hooks on non-leaf functions (e.g. `a.grad_fn` here) already work without this PR.

> 2) post hooks on accumulate grad

I tried the example code. `t.grad_fn` is `None` in this example. Any suggestion?

@pritamdamania87

Reply:
> We can test that, but want to make sure we are on the same page. Post hooks on non-leaf functions (e.g. a.grad_fn here) already work without this PR.

Yes, I meant to keep the current test.

> I tried the example code. t.grad_fn is None in this example. Any suggestion?

Yes, thats why I had to use `a.grad_fn.next_functions[0][0]` to get accumulate_grad. Seems like you would have to do something similar. I'm not sure why t.grad_fn is None though. cc @albanD Is this expected for accumulate grad?

Reply:
> Is this expected for accumulate grad?

Yes, this is expected. In particular for the leaf definition.
The AccumulateGrad Node is a trick we use to accumulate the gradient, but the Tensor is a leaf that requires grad (so has no grad_fn) as it was not created in a differentiable manner.